src/explore-education-statistics-admin/src/pages/release/edit-release/content/ReleaseContext.tsx

```diff
@@ -0,0 +1,237 @@
+import produce from 'immer';
+import {
+  EditableContentBlock,
+  ExtendedComment,
+  EditableRelease,
+} from '@admin/services/publicationService';
+import { DataBlock } from '@common/services/dataBlockService';
+import { ContentSection } from '@common/services/publicationService';
+import React, { createContext, useContext, useReducer } from 'react';
+import ReleaseDispatchAction from './actions';
+
+type Dispatch = (action: ReleaseDispatchAction) => void;
+type State = {
+  release: EditableRelease | undefined;
+  canUpdateRelease: boolean;
+  availableDataBlocks: DataBlock[];
+  unresolvedComments: ExtendedComment[];
+};
+type ReleaseProviderProps = { children: React.ReactNode };
+
+const ReleaseStateContext = createContext<State | undefined>(undefined);
+const ReleaseDispatchContext = createContext<Dispatch | undefined>(undefined);
+
+export function releaseReducer(state: State, action: ReleaseDispatchAction) {
+  switch (action.type) {
+    case 'CLEAR_STATE':
+      return {
+        release: undefined,
+        canUpdateRelease: false,
+        availableDataBlocks: [],
+        unresolvedComments: [],
+      };
+    case 'SET_STATE':
+    case 'SET_AVAILABLE_DATABLOCKS':
+      return { ...state, ...action.payload };
+    case 'REMOVE_BLOCK_FROM_SECTION':
+      return produce<State>(state, draft => {
+        const { sectionId, blockId, sectionKey } = action.payload.meta;
+        if (
+          draft.release === undefined ||
+          typeof draft.release[sectionKey] !== 'object'
+        ) {
+          throw new Error('REMOVE_BLOCK_FROM_SECTION: failed');
+        } else if (sectionKey === 'content') {
+          draft.release[sectionKey] = draft.release[sectionKey].map(section => {
+            if (section.id === sectionId) {
+              return {
+                ...section,
+                content: section.content?.filter(
+                  contentBlock => (contentBlock.id as string) !== blockId,
+                ),
+              } as ContentSection<EditableContentBlock>;
+            }
+            return section;
+          });
+        } else {
+          (draft.release[sectionKey] as ContentSection<
+            EditableContentBlock
+          >).content = (draft.release[sectionKey] as ContentSection<
+            EditableContentBlock
+          >).content?.filter(
+            contentBlock => (contentBlock.id as string) !== blockId,
+          );
+        }
+      });
+    case 'UPDATE_BLOCK_FROM_SECTION': {
+      return produce<State>(state, draft => {
+        const { block, meta } = action.payload;
+        const { sectionId, blockId, sectionKey } = meta;
+        if (
+          draft.release === undefined ||
+          typeof draft.release[sectionKey] !== 'object'
+        ) {
+          throw new Error('UPDATE_BLOCK_FROM_SECTION: failed');
+        } else if (sectionKey === 'content') {
+          draft.release[sectionKey] = draft.release[sectionKey].map(section => {
```

Comment:
Take the suggestions from above to write the rest of this branch in a more Immer idiomatic way.